src/core/dyad_core.c

```diff
@@ -711,9 +730,75 @@ dyad_rc_t dyad_produce (dyad_ctx_t* ctx, const char* fname)
     return dyad_commit (ctx, fname);
 }
 
+// DYAD_CORE_FUNC_MODS dyad_rc_t dyad_kvs_lookup (const dyad_ctx_t* ctx,
+//                                                const char* restrict kvs_topic,
+//                                                uint32_t* owner_rank,
+//                                                flux_future_t** f)
+dyad_rc_t dyad_get_metadata (dyad_ctx_t* ctx,
+                             const char* fname,
+                             bool should_wait,
+                             dyad_metadata_t** mdata)
+{
+    dyad_rc_t rc = DYAD_RC_OK;
+    const size_t topic_len = PATH_MAX;
+    char topic[PATH_MAX + 1];
+    char upath[PATH_MAX];
+    memset (topic, 0, topic_len + 1);
+    memset (topic, '\0', topic_len + 1);
+    memset (upath, 0, PATH_MAX);
+    // Extract the path to the file specified by fname relative to the
+    // producer-managed path
+    // This relative path will be stored in upath
+    DYAD_LOG_INFO (ctx, "Obtaining file path relative to consumer directory: %s", upath);
+    if (!cmp_canonical_path_prefix (ctx->cons_managed_path, fname, upath, PATH_MAX)) {
+        DYAD_LOG_INFO (ctx, "%s is not in the Consumer's managed path\n", fname);
+        rc = DYAD_RC_UNTRACKED;
+        goto get_metadata_done;
+    }
+    DYAD_LOG_INFO (ctx, "Generating KVS key: %s", topic);
+    gen_path_key (upath, topic, topic_len, ctx->key_depth, ctx->key_bins);
+    rc = dyad_kvs_read (ctx, topic, should_wait, mdata);
+    if (DYAD_IS_ERROR (rc)) {
+        DYAD_LOG_ERR (ctx, "Could not read data from the KVS");
+        goto get_metadata_done;
+    }
+    free ((*mdata)->fpath);
+    size_t fname_len = strlen (fname);
+    (*mdata)->fpath = (char*)malloc (fname_len + 1);
+    if ((*mdata)->fpath == NULL) {
+        DYAD_LOG_ERR (ctx, "Could not allocate memory for fpath");
+        rc = DYAD_RC_SYSFAIL;
+        goto get_metadata_done;
+    }
+    memset ((*mdata)->fpath, '\0', fname_len + 1);
+    strncpy ((*mdata)->fpath, fname, fname_len);
+    rc = DYAD_RC_OK;
+
+get_metadata_done:
+    if (DYAD_IS_ERROR (rc) && mdata != NULL && *mdata != NULL) {
+        if ((*mdata)->fpath != NULL)
+            free ((*mdata)->fpath);
+        free (*mdata);
+        *mdata = NULL;
+    }
+    return rc;
+}
+
+dyad_rc_t dyad_free_metadata (dyad_metadata_t* mdata)
```

Comment:
mdata needs to be pointer of pointer to be updated to NULL at the end

Reply:
Good catch. I'll change the pointer type.